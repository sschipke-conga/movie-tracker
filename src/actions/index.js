export const setMovies = (movies) => ({
  type: 'SET_MOVIES',
  movies,
});

export const setUpcomingMovies = (upcomingMovies) => ({
  type: 'SET_UPCOMING_MOVIES',
  upcomingMovies,
});

export const setUser = (user) => ({
  type: 'SET_USER',
  user,
});

export const setFavorites = (favorites) => ({
  type: 'SET_FAVORITES',
  favorites,
});
