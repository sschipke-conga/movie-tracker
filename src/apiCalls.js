export const getMovies = async () => {
  let res = await fetch('https://api.themoviedb.org/3/discover/movie?sort_by=popularity.desc&api_key=9e1950f701d0af712babdede31811e9e')
  let data = await res.json()
  return data.results
}