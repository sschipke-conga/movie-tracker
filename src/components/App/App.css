*{
  box-sizing: border-box;
  margin:0;
  padding:0;
  background-color:black;
  color:white;
  font-family: 'Istok Web', sans-serif;
}


.App {
  text-align: center;
}

.App > section {
  margin-top:120px;
}
