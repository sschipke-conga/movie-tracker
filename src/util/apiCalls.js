export const getMovies = async () => {
  let res = await fetch('https://api.themoviedb.org/3/movie/now_playing?api_key=9e1950f701d0af712babdede31811e9e&language=en-US')
  let data = await res.json()
  return cleanMovieData(data.results)
}

const cleanMovieData = data => data.map(movie => ({
    poster_path: movie.poster_path,
    title: movie.title,
    release_date: movie.release_date,
    vote_average: movie.vote_average,
    overview: movie.overview,
    movie_id: movie.id
  }))

export const getUpcomingMovies = async () => {
  let res = await fetch('https://api.themoviedb.org/3/movie/upcoming?api_key=9e1950f701d0af712babdede31811e9e&language=en-US')
  let data = await res.json()
  return cleanMovieData(data.results)
}

export const createNewUser = async user => {
  let url = 'http://localhost:3001/api/v1/users'
  let options = {
      method: 'POST',
      body: JSON.stringify(user),
        headers: {
      'Content-Type': 'application/json'
    }
  }
  let res = await fetch(url, options);

  if (!res.ok) {
    if(res.status===500) {
      throw Error('This email has already been used')
    }
  }
  return res.json();
}

export const logInUser = async user => {
  let url = 'http://localhost:3001/api/v1/login';
  let options = {
    method: 'POST',
    body: JSON.stringify(user),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  let res = await fetch(url, options)

  if (!res.ok) {
    if (res.status === 401) {
      throw Error('email or password is incorrect')
    }
  }
  return res.json();
}

export const getUserFavorites = async userID => {
  let url = `http://localhost:3001/api/v1/users/${userID}/moviefavorites`;
  let res = await fetch(url);
  let parsedRes = await res.json();
  return parsedRes.favorites;
}

export const postFavorite = async (userId, movie) => {
  console.log('post', userId, movie)
  let url = `http://localhost:3001/api/v1/users/${userId}/moviefavorites`;
  let options = {
    method: 'POST',
    body: JSON.stringify(movie),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  let res = await fetch(url, options);
  if (!res.ok) {
    throw Error('Could not add favorite movie.')
  }
}


export const deleteFavorite = async (userId, movie_id) => {
  console.log('delete', userId, movie_id)
  let url = `http://localhost:3001/api/v1/users/${userId}/moviefavorites/${movie_id}`
  let options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  }
  let res = await fetch(url, options);
  if (!res.ok) {
      throw Error('Could not delete favorite.')
  }

}
