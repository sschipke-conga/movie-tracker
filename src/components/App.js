
import React, { Component } from 'react';
import { Route } from 'react-router-dom';


//Importing Other Componets
import Login from './Login/Login';
import Nav from './Nav/Nav';
import Favorites from './Favorites/Favorites';
import Main from './Main/Main'
import MoviePage from './MoviePage/MoviePage';
import {getMovies, getUpcomingMovies} from '../util/apiCalls'
import {setMovies, setUpcomingMovies } from '../actions'
import MovieList from './MovieList/MovieList'
import './App.css';
import { connect } from 'react-redux';


export class App extends Component { 

  async componentDidMount() {
    const {setMovies, setUpcomingMovies} = this.props;
    try {
      const data = await getMovies();
      setMovies(data)
    } catch(error) {
      console.log(error)
    }

    try {
      const data = await getUpcomingMovies();
      setUpcomingMovies(data)
    } catch(error) {
      console.log (error)
    }
   


  }

  render = () => {
    return (
      <div className="App">
        <Route exact path='/login' render={ (props)=> <Login {...props}/>} />
        <Route path='/' render={ () => <Nav /> } />
        <Route exact path='/' render={ () => <Main /> } />
        <Route exact path='/favorites' render={ () => <Favorites /> } />
        <Route exact path='/movie/:id' render={ () => <MoviePage /> } />
        <Route exact path='/upcoming' render={ () => <MovieList movies={this.props.upcomingMovies}/> } />
        <Route exact path='/now_playing' render={ () => <MovieList movies={this.props.movies}/> } />
      </div> 
    );
  }
}

export const mapStateToProps = state => ({
  movies: state.movies,
  upcomingMovies: state.upcomingMovies,
})

export const mapDispatchToProps = dispatch => ({
  setMovies: movies => dispatch(setMovies(movies)),
  setUpcomingMovies: (upcomingMovies) => dispatch(setUpcomingMovies(upcomingMovies))
})

export default connect(mapStateToProps, mapDispatchToProps) (App)
