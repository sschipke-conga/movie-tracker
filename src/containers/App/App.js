
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Route } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import Login from '../Login/Login';
import Nav from '../Nav/Nav';
import Favorites from '../Favorites/Favorites';
import Main from '../Main/Main';
import MoviePage from '../../components/MoviePage/MoviePage';
import {getMovies, getUpcomingMovies, getUserFavorites, deleteFavorite, postFavorite, logInUser} from '../../util/apiCalls';
import { setMovies, setUpcomingMovies, setFavorites, setUser } from '../../actions';
import MovieList from '../../components/MovieList/MovieList';
import './App.css';

export class App extends Component {
  async componentDidMount() {
    const {
      setMovies,
      setUpcomingMovies,
      setFavorites,
      setUser,
      user
    } = this.props;
    try {
      const data = await getMovies();
      setMovies(data);
    } catch (error) {
      console.log(error);
    }

    try {
      const data = await getUpcomingMovies();
      setUpcomingMovies(data);
    } catch (error) {
      console.log (error)
    };

    if(localStorage.getItem("user")) {
      console.log(JSON.parse(localStorage.getItem("user")))
      const savedUser = JSON.parse(localStorage.getItem("user"));
      const {email, password} = savedUser
      try {
        let userParams = {email, password}
        let currentUser = await logInUser(userParams);
        let userFavorites = await getUserFavorites(currentUser.id);
        setFavorites(userFavorites);
        setUser(currentUser);
      } catch({message}) {
        console.log(message)
      }
    }

    if(user.name) {
      try {
        const userFavs = await getUserFavorites(user.id);
        setFavorites(userFavs);
      } catch (error) {
        console.log(error);
      }
    }
  }

  toggleFavorites = (e, movie) => {
    e.preventDefault();
    const userID = this.props.user.id;
    if (this.props.favorites.map((fav) => fav.title).includes(movie.title)) {
      this.removeFavorite(userID, movie.movie_id);
    } else {
      this.addFavorite(userID, movie);
    }
  }

  removeFavorite = async (userID,movieID) => {
    try {
      await deleteFavorite(userID, movieID);
      const updatedFavorites = await getUserFavorites(userID);
      this.props.setFavorites(updatedFavorites);
    } catch ({ message }) {
      console.log(message);
    }
  }

  addFavorite = async (userID, movie) => {
    try {
      await postFavorite(userID, movie);
      const currentFavorites = await getUserFavorites(userID);
      this.props.setFavorites(currentFavorites);
    } catch({ message }) {
      console.log(message);
    }
  };

  logOut = () => {
    localStorage.clear()
  }



  render = () => {
    return (
      <div className="App">
        <Route exact path='/login' render={ (props)=> <Login {...props}/>} />
        <Route path='/' render={ () => <Nav logOut={this.logOut} /> } />
        <Route exact path='/' render={() => <Main toggleFavorites={this.toggleFavorites} /> } />
        <Route exact path='/favorites' render={ () => <Favorites toggleFavorites={this.toggleFavorites} movies={this.props.favorites} /> } />
        <Route exact path='/movie/:movie_id' render={ ({match}) => {
          let allMovies = [...this.props.movies, ...this.props.upcomingMovies, ...this.props.favorites];
          let currentMovie = allMovies.find(movie => movie.movie_id === parseInt(match.params.movie_id))
          return (<MoviePage {...currentMovie}/>)
        }}/>
        <Route exact path='/upcoming' render={() => <MovieList toggleFavorites={this.toggleFavorites} movies={this.props.upcomingMovies}/> } />
        <Route exact path='/now_playing' render={() => <MovieList toggleFavorites={this.toggleFavorites}movies={this.props.movies}/> } />
      </div> 
    );
  };
}

export const mapStateToProps = (state) => ({
  movies: state.movies,
  upcomingMovies: state.upcomingMovies,
  favorites: state.favorites,
  user: state.user,
});

export const mapDispatchToProps = dispatch => (bindActionCreators({
  setMovies, setUpcomingMovies, setFavorites, setUser
}, dispatch));

export default connect(mapStateToProps, mapDispatchToProps) (App);

App.propTypes = {
  movies: PropTypes.array,
  upcomingMovies: PropTypes.array,
  favorites: PropTypes.array,
  user: PropTypes.object,
  setMovies: PropTypes.func.isRequired,
  setUpcomingMovies: PropTypes.func.isRequired,
  setFavorites: PropTypes.func.isRequired
}
