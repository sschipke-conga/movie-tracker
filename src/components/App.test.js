import React from 'react';
import ReactDOM from 'react-dom';
import { App, mapStateToProps, mapDispatchToProps} from './App';
import { shallow } from 'enzyme';
import { getMovies, getUpcomingMovies } from '../util/apiCalls'
import { setMovies, setUpcomingMovies } from '../actions'


jest.mock('../util/apiCalls.js')

describe('App',() => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<App />)
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
})

describe('App REDUX test',() => {
  let mockMovies, mockUpcomingMovies;
  beforeEach(() => {
    mockMovies = [
      {
        "poster_path": "/udDclJoHjfjb8Ekgsd4FDteOkCU.jpg",
        "id": 475557,
        "title": "Joker",
        "vote_average": 8.6,
        "overview": "During the 1980s, a failed stand-up comedian is driven insane and turns to a life of crime and chaos in Gotham City while becoming an infamous psychopathic crime figure.",
        "release_date": "2019-10-04"
      },
      {
        "poster_path": "/tBuabjEqxzoUBHfbyNbd8ulgy5j.jpg",
        "id": 420809,
        "title": "Maleficent: Mistress of Evil",
        "vote_average": 7.1,
        "overview": "Maleficent and her goddaughter Aurora begin to question the complex family ties that bind them as they are pulled in different directions by impending nuptials, unexpected allies, and dark new forces at play.",
        "release_date": "2019-10-18"
      }
    ];
    mockUpcomingMovies = [
      {
        "poster_path": "/vqzNJRH4YyquRiWxCCOH0aXggHI.jpg",
        "id": 290859,
        "title": "Terminator: Dark Fate",
        "vote_average": 7.4,
        "overview": "More than two decades have passed since Sarah Connor prevented Judgment Day, changed the future, and re-wrote the fate of the human race. Dani Ramos is living a simple life in Mexico City with her brother and father when a highly advanced and deadly new Terminator – a Rev-9 – travels back through time to hunt and kill her. Dani's survival depends on her joining forces with two warriors: Grace, an enhanced super-soldier from the future, and a battle-hardened Sarah Connor. As the Rev-9 ruthlessly destroys everything and everyone in its path on the hunt for Dani, the three are led to a T-800 from Sarah’s past that may be their last best hope.",
        "release_date": "2019-11-01"
      },
      {
        "poster_path": "/pIcV8XXIIvJCbtPoxF9qHMKdRr2.jpg",
        "id": 338967,
        "title": "Zombieland: Double Tap",
        "vote_average": 7.3,
        "overview": "The group will face a new zombie threat as a new breed of zombie has developed. This new super-zombie type is faster, bigger, and stronger than the previous strain of zombies and harder to kill. These super-zombies have started grouping up into a horde going from city to city leaving a path of destruction behind them.",
        "release_date": "2019-10-18"
      }
    ];
    
  })

  it('mapStateToProps give all the todos in state',() => {
    //Setup

    const mockState = { 
      movies: mockMovies,
      upcomingMovies: mockUpcomingMovies,
      favotites:[],
      user:{id:1}
      };

    const expected = {
      movies: mockState.movies,
      upcomingMovies: mockState.upcomingMovies
    };
    //Execution
    const mappedProps = mapStateToProps(mockState)
    //Assertion
    expect(mappedProps).toEqual(expected)
  });

  it('calls dispatch with setMovies action when setMovies is called', () => {
    // Setup
    const mockDispatch = jest.fn();
    const actionToDispatch = setMovies('SET_MOVIES', mockMovies);
    // Execution
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setMovies('SET_MOVIES', mockMovies);
    // Expectaion
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

});