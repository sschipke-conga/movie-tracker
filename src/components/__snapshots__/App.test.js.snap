// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="App"
>
  <Route
    exact={true}
    path="/login"
    render={[Function]}
  />
  <Route
    path="/"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/favorites"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/movie/:movie_id"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/upcoming"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/now_playing"
    render={[Function]}
  />
</div>
`;
